app/src/main/java/xyz/block/gosling/features/agent/providers/OpenRouterProviderHandler.kt

```diff
@@ -0,0 +1,277 @@
+
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import org.json.JSONObject
+import xyz.block.gosling.features.agent.*
+import java.lang.reflect.Method
+import java.util.*
+
+/**
+ * OpenRouter provider handler for accessing multiple model providers through a single API.
+ * Handles models from Anthropic, Meta, Mistral, Cohere, and many others via OpenRouter.
+ */
+class OpenRouterProviderHandler : LLMProviderHandler {
+    
+    private val jsonFormat = Json {
+        prettyPrint = true
+        encodeDefaults = true
+        ignoreUnknownKeys = true
+    }
+    
+    override fun createToolDefinitions(toolMethods: List<Method>): SerializableToolDefinitions {
+        val toolDefinitions = toolMethods.mapNotNull { method ->
+            val tool = method.getAnnotation(Tool::class.java) ?: return@mapNotNull null
+
+            // Create tool parameters for OpenRouter (OpenAI-compatible format)
+            val toolParameters = ToolParametersObject(
+                properties = tool.parameters.associate { param ->
+                    param.name to ToolParameter(
+                        type = param.type,
+                        description = param.description
+                    )
+                },
+                required = tool.parameters
+                    .filter { it.required }
+                    .map { it.name }
+            )
+
+            ToolDefinition(
+                function = ToolFunctionDefinition(
+                    name = tool.name,
+                    description = tool.description,
+                    parameters = toolParameters
+                )
+            )
+        }
+        
+        return SerializableToolDefinitions.OpenAITools(toolDefinitions)
+    }
+    
+    override fun createRequest(
+        modelIdentifier: String,
+        messages: List<Message>,
+        tools: SerializableToolDefinitions,
+        apiKey: String?
+    ): String {
+        val toolDefinitions = when (tools) {
+            is SerializableToolDefinitions.OpenAITools -> tools.definitions
+            else -> emptyList()
+        }
+
+        // OpenRouter uses OpenAI-compatible format
+        // Note: Not all models support tool calling - check OpenRouter docs for model-specific capabilities
+        
+        // Convert messages to OpenAI format, handling tool messages specially
+        val openAIMessages = messages.map { message ->
+            when (message.role) {
+                "tool" -> {
+                    // Tool messages need special handling for OpenRouter
+                    val content = message.content?.firstOrNull()
+                    val textContent = if (content is Content.Text) content.text else content.toString()
+                    
+                    // Create a tool message using JSON directly
+                    org.json.JSONObject().apply {
+                        put("role", "tool")
+                        put("content", textContent)
```

Comment:
Same as the above, but here I'm a bit more concerned because we are in a loop.

```suggestion
                    // Create a tool message using JSON directly
                    org.json.JSONObject().apply {
                        put("role", "tool")
                        put("content",  if (content is Content.Text) content.text else content.toString())
```

Reply:
Nice! I went ahead and included this in my latest push, thanks @bricefriha :)

Reply:
didn't see you had two of these, just updated this in my following push as well. Thanks for the callouts! :)